File: ios-badania-app/BadaniaApp/BadaniaApp/ViewModels/BadaniaViewModel.swift
//
//  BadaniaViewModel.swift
//  BadaniaApp
//
//  ViewModel zarządzający listą wszystkich badań i wyszukiwaniem
//

import Foundation
import Combine

@MainActor
class BadaniaViewModel: ObservableObject {
    @Published var wszystkieBadania: [Badanie] = []
    @Published var filtrowaneBadania: [Badanie] = []
    @Published var searchText: String = ""
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        setupSearchFilter()
        loadBadania()
    }
    
    /// Wczytuje badania z pliku CSV
    func loadBadania() {
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                let badania = try CSVParser.parseCSV()
                await MainActor.run {
                    self.wszystkieBadania = badania.sorted { $0.nazwa.lowercased() < $1.nazwa.lowercased() }
                    // Zaktualizuj filtrowane badania po załadowaniu
                    self.filtrowaneBadania = self.filterBadania(searchText: self.searchText)
                    self.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.isLoading = false
                    self.errorMessage = "Błąd podczas wczytywania danych: \(error.localizedDescription)"
                }
            }
        }
    }
    
    /// Konfiguruje filtrowanie w czasie rzeczywistym podczas wpisywania
    private func setupSearchFilter() {
        $searchText
            .debounce(for: .milliseconds(300), scheduler: RunLoop.main)
            .map { [weak self] searchText in
                self?.filterBadania(searchText: searchText) ?? []
            }
            .assign(to: &$filtrowaneBadania)
    }
    
    /// Filtruje badania na podstawie tekstu wyszukiwania (case-insensitive)
    private func filterBadania(searchText: String) -> [Badanie] {
        if searchText.isEmpty {
            return wszystkieBadania
        }
        
        let lowercasedSearch = searchText.lowercased()
        return wszystkieBadania.filter { badanie in
            badanie.nazwa.lowercased().contains(lowercasedSearch) ||
            badanie.kod.lowercased().contains(lowercasedSearch)
        }
    }
}

File: android-badania-app/app/src/main/java/com/decentcode/badaniaapp/viewmodels/BadaniaViewModel.kt
package com.decentcode.badaniaapp.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.decentcode.badaniaapp.models.Badanie
import com.decentcode.badaniaapp.models.CSVParser
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * ViewModel zarządzający listą wszystkich badań i wyszukiwaniem
 */
class BadaniaViewModel(application: Application) : AndroidViewModel(application) {
    private val _wszystkieBadania = MutableStateFlow<List<Badanie>>(emptyList())
    val wszystkieBadania: StateFlow<List<Badanie>> = _wszystkieBadania.asStateFlow()
    
    private val _filtrowaneBadania = MutableStateFlow<List<Badanie>>(emptyList())
    val filtrowaneBadania: StateFlow<List<Badanie>> = _filtrowaneBadania.asStateFlow()
    
    private val _searchText = MutableStateFlow("")
    val searchText: StateFlow<String> = _searchText.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    private var searchJob: Job? = null
    
    init {
        setupSearchFilter()
        loadBadania()
    }
    
    /**
     * Wczytuje badania z pliku CSV
     */
    fun loadBadania() {
        _isLoading.value = true
        _errorMessage.value = null
        
        viewModelScope.launch {
            CSVParser.parseCSV(getApplication()).fold(
                onSuccess = { badania ->
                    _wszystkieBadania.value = badania
                    _filtrowaneBadania.value = filterBadania(_searchText.value)
                    _isLoading.value = false
                },
                onFailure = { error ->
                    _isLoading.value = false
                    _errorMessage.value = "Błąd podczas wczytywania danych: ${error.message}"
                }
            )
        }
    }
    
    /**
     * Aktualizuje tekst wyszukiwania
     */
    fun updateSearchText(text: String) {
        _searchText.value = text
    }
    
    /**
     * Konfiguruje filtrowanie w czasie rzeczywistym podczas wpisywania z debounce
     */
    private fun setupSearchFilter() {
        // Debounce będzie obsługiwany przez updateSearchText
    }
    
    /**
     * Filtruje badania na podstawie tekstu wyszukiwania (case-insensitive)
     */
    private fun filterBadania(searchText: String): List<Badanie> {
        if (searchText.isEmpty()) {
            return _wszystkieBadania.value
        }
        
        val lowercasedSearch = searchText.lowercase()
        return _wszystkieBadania.value.filter { badanie ->
            badanie.nazwa.lowercase().contains(lowercasedSearch) ||
            badanie.kod.lowercase().contains(lowercasedSearch)
        }
    }
    
    /**
     * Aktualizuje filtrowane badania z debounce
     */
    fun performSearch() {
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            delay(300) // Debounce 300ms
            _filtrowaneBadania.value = filterBadania(_searchText.value)
        }
    }
}

